datasource/virtualmachine/data.go

```diff
@@ -0,0 +1,193 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+//go:generate packer-sdc struct-markdown
+//go:generate packer-sdc mapstructure-to-hcl2 -type Config,Tag,DatasourceOutput
+package virtualmachine
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/hashicorp/hcl/v2/hcldec"
+	"github.com/hashicorp/packer-plugin-sdk/common"
+	"github.com/hashicorp/packer-plugin-sdk/hcl2helper"
+	packersdk "github.com/hashicorp/packer-plugin-sdk/packer"
+	"github.com/hashicorp/packer-plugin-sdk/template/config"
+	vsphere "github.com/hashicorp/packer-plugin-vsphere/builder/vsphere/common"
+	"github.com/hashicorp/packer-plugin-vsphere/datasource/common/driver"
+	"github.com/zclconf/go-cty/cty"
+)
+
+// HCL Example:
+//
+// ```hcl
+//
+//	tag {
+//	  category = "team"
+//	  name = "operations"
+//	}
+//	tag {
+//	  category = "sla"
+//	  name = "gold"
+//	}
+//
+// ```
+type Tag struct {
+	// Name of the tag added to virtual machine which must pass the `tag`
+	// filter.
+	Name string `mapstructure:"name" required:"true"`
+	// Name of the tag category that contains the tag.
+	//
+	// -> **Note:** Both `name` and `category` must be specified in the `tag`
+	// filter.
+	Category string `mapstructure:"category" required:"true"`
+}
+
+type Config struct {
+	common.PackerConfig   `mapstructure:",squash"`
+	vsphere.ConnectConfig `mapstructure:",squash"`
+
+	// Basic filter with glob support (e.g. `ubuntu_basic*`). Defaults to `*`.
+	// Using strict globs will not reduce execution time because vSphere API
+	// returns the full inventory. But can be used for better readability over
+	// regular expressions.
+	Name string `mapstructure:"name"`
+	// Extended name filter with regular expressions support
+	// (e.g. `ubuntu[-_]basic[0-9]*`). Default is empty. The match of the
+	// regular expression is checked by substring. Use `^` and `$` to define a
+	// full string. For example, the `^[^_]+$` filter will search names
+	// without any underscores. The expression must use
+	// [Go Regex Syntax](https://pkg.go.dev/regexp/syntax).
+	NameRegex string `mapstructure:"name_regex"`
+	// Filter to return only objects that are virtual machine templates.
+	// Defaults to `false` and returns all virtual machines.
+	Template bool `mapstructure:"template"`
+	// Filter to search virtual machines only on the specified ESX host.
+	Host string `mapstructure:"host"`
+	// Filter to return only that virtual machines that have attached all
+	// specifies tags. Specify one or more `tag` blocks to define list of tags
+	//  for the filter. Multiple blocks can be created programmatically in HCL2 with the [`dynamic_block`](/packer/docs/templates/hcl_templates/expressions#dynamic-blocks).
```

Comment:
Perhaps something like this would be cleaner?

```suggestion
	Host string `mapstructure:"host"`
	// Filter to return only that virtual machines that have attached all
	// specifies tags. Specify one or more `tag` blocks to define the tags
	// for the filter.
	//
	// HCL Example:
	//
	// ```hcl
	//	tag {
	//	  category = "team"
	//	  name = "operations"
	//	}
	//	tag {
	//	  category = "sla"
	//	  name = "gold"
	//	}
	// ```
	//
	// You can dynamically construct repeatable nested blocks
	// for `tag` using [`dynamic_block`](/packer/docs/templates/hcl_templates/expressions#dynamic-blocks).
	//
	// HCL Example:
	//
	// ```hcl
	// locals {
	//   standard_tags = [
	//     {
	//       category = "team"
	//       name     = "operations"
	//     },
	//     {
	//       category = "sla"
	//       name     = "gold"
	//     }
	//   ]
	// }
	//
	// source "vsphere" "example" {
	//   # ...
	//   dynamic "tag" {
	//     for_each = local.standard_tags
	//     content {
	//       key   = tag.value["category"]
	//       value = tag.value["name"]
	//     }
	//   }
	// }
	// ```
```

Reply:
Oh, now I understand. Maybe I should leave the general description dry and describe the dynamic block in the tag description section?

Reply:
I think that the above suggestion might render rather nicely in the docs so it's easier to follow from a simple example to one with dynamic blocks right after.

Reply:
I mean that it's clearer to me when the list of available options contains only the list of options. And examples are described rigth after the list or the option, which needs example, has an in-doc link to the descrption somewhere later in the page.
I would like to have documentation to look like this:
<img src="https://github.com/user-attachments/assets/d1d63f12-2d27-41d0-ac7c-8b95450d55ae" height="300">

Reply:
That would work too. :)

Reply:
It seems that Packer HCL still have no support for `dynamic` blocks in datasources. I have tried multiple times with different configs, tried to find some affirmtaions in code of Packer itself that it supports `dynamic` in datasources, but no luck. =(

Reply:
Bummer!

Reply:
Updated the suggestion.